### ISSUE TITLE
Qwen3‑VL on‑policy KD silently trains "blind" with sglang 0.5.9 (vision weights not loaded; fixed by   sglang#19333)

### ISSUE THREAD

**Issue by Zhaolutuan** [songmzhang/KDFlow#9]
### Summary
  When running on-policy VL distillation (Qwen3-VL-8B → Qwen3-VL-2B, Geo3K) with
  **sglang 0.5.9**, the model never actually sees the images, yet training shows
  **no error and looks completely healthy** — this is a silent failure that is very
  easy to mistake for a working run.

  ### Environment
  - KDFlow on-policy KD (`train_kd_on_policy`), `kd_loss_fn=rkl`, `kd_ratio=1.0`
  - teacher Qwen3-VL-8B-Instruct, student Qwen3-VL-2B-Instruct, dataset Geo3K (image+text)
  - sglang **0.5.9**, transformers 4.57.1, torch 2.9.1+cu128, 8×A100

  ### Symptoms (all "look fine")
  - `loss` / `kd_loss` converge smoothly, `grad_norm` normal, **no NaN, no exception**
  - but the rollout/eval generations show the model can't see the figure, e.g.:
    - `"The image you provided is blank and contains no content..."`
    - `"...the image isn't here. I can't see the green figure..."`
  - `eval` accuracy stays at chance (~1% on geometry) — i.e. pure text guessing.

  Because loss decreases regardless (student just mimics the teacher's *equally blind*
  text distribution), the run looks successful. Nothing surfaces the problem.

  ### Root cause (upstream, in sglang — not KDFlow)
  sglang 0.5.9 fails to load the **Qwen3-VL vision tower weights** because of a
  parameter-name prefix mismatch (`model.visual.*` in the checkpoint vs `visual.*`
  expected by sglang). The image placeholder tokens *are* expanded (the prompt
  sequence grows), but the vision features filling them are from **uninitialized
  weights**, so the LM effectively gets no image.

  Fixed upstream by **sgl-project/sglang#19333** ("fix qwen3_vl visual module loading"):
  https://github.com/sgl-project/sglang/pull/19333

  ### Minimal reproduction (no KDFlow involved)
  Launch a stock sglang 0.5.9 server for Qwen3-VL and send any image via the
  OpenAI chat-completions API — the model does not describe the image:

  ```bash
  python -m sglang.launch_server --model-path <Qwen3-VL-8B-Instruct> \
    --dtype bfloat16 --trust-remote-code --port 30000 --served-model-name qwen3vl8b
  # POST /v1/chat/completions with an image_url -> reply ignores the image

  Fix / workaround

  Apply the change from sglang#19333 to python/sglang/srt/models/qwen3_vl.py
  (the visual-module weight-name remap), or upgrade sglang to a release that
  includes that PR. It's a pure-Python change — just restart the process, no
  recompilation of sgl-kernel needed. After this, the vision weights load and
  the model sees the images.

  Suggestion for KDFlow

  Since this is a silent, no-error failure that wastes full training runs, it'd
  help to (a) pin a known-good sglang version in requirements, and/or (b) add a
  note in the VL examples README pointing at sglang#19333.

**Comment by songmzhang** [comment]
Thanks for pointing this out! This is a very valuable issue for the distillation of VL models. I'll update the requirements soon and mention this issue in README.md.

**Comment by rehan243** [comment]
oh interesting, ran into almost the same silent failure a few weeks back — not gonna lie, it was brutal to debug since loss and gradients were totally normal. We were distilling a VL model (not Qwen3-VL, but similar setup) and the student just kept parroting text despite eval images being clearly part of the prompt. Turned out, vision weights never loaded due to a busted key prefix in the state_dict, but nothing ever threw an error. The thing is, transformers 4.57.1 has better logging for missing modules, but sglang 0.5.9 just quietly skips anything it can't match if the names don't align. I remember our eval generations saying stuff like “I can’t see your image” — exactly like your `"The image you provided is blank..."` symptom. We only caught it because eval accuracy tanked and our prod run was stuck at random guessing for image tasks (like ~1% on geometry, same as you). Honestly, if the teacher’s vision weights aren’t loaded, the student’s loss will still drop because it’s just mimicking whatever weird text-only distribution the teacher outputs. Took us a few tries to get the prefix right: needed to map `model.visual.*` to `visual.*` manually before loading. The docs don’t mention this at all.

**Comment by nostrebor** [comment]
fixed on main already, should be closed

PR #19333 landed in sglang 0.5.10 and current reqs specify >=0.5.10

https://github.com/songmzhang/KDFlow/blob/main/requirements.txt
